native/c/zds.cpp

```diff
@@ -992,119 +1470,166 @@ int zds_list_data_sets(ZDS *zds, string dsn, vector<ZDSEntry> &attributes)
+          if (actual_len > 0 && actual_len <= (*field_len - 2))
+          {
+            entry.dataclass = string((char *)(data + 2), actual_len);
+          }
+        }
 
-        char non_vsam_attribute = 0xFF;
+        // Parse MGMTCLAS field (8 bytes)
+        data += *field_len;
+        field_len++;
+        entry.mgmtclass = "";
+        if (*field_len > 2)
+        {
+          // First 2 bytes are a length prefix, extract actual length
+          uint16_t actual_len = (static_cast<unsigned char>(data[0]) << 8) | static_cast<unsigned char>(data[1]);
+          if (actual_len > 0 && actual_len <= (*field_len - 2))
+          {
+            entry.mgmtclass = string((char *)(data + 2), actual_len);
+          }
+        }
 
-        if (*field_len > 0)
+        // Parse STORCLAS field (8 bytes)
+        data += *field_len;
+        field_len++;
+        entry.storclass = "";
+        if (*field_len > 2)
         {
-          non_vsam_attribute = *(char *)data;
+          // First 2 bytes are a length prefix, extract actual length
+          uint16_t actual_len = (static_cast<unsigned char>(data[0]) << 8) | static_cast<unsigned char>(data[1]);
+          if (actual_len > 0 && actual_len <= (*field_len - 2))
+          {
+            entry.storclass = string((char *)(data + 2), actual_len);
+          }
+        }
 
-#define SIMPLE_NON_VSAM_DATA_SET 0X00
-#define EXTENDED_PARTITIONED_DATA_SET 'L'
+        // Parse DEVTYP field (4-byte UCB device type)
+        data += *field_len;
+        field_len++;
+        entry.devtype = "";
+        if (*field_len >= 4)
+        {
+          uint32_t devtyp = (static_cast<unsigned char>(data[0]) << 24) |
+                            (static_cast<unsigned char>(data[1]) << 16) |
+                            (static_cast<unsigned char>(data[2]) << 8) |
+                            static_cast<unsigned char>(data[3]);
 
-          if (EXTENDED_PARTITIONED_DATA_SET == non_vsam_attribute)
+          // If devtyp is all zeros, default to 3390 (most common modern DASD)
+          if (devtyp == 0x00000000)
           {
-            entry.dsorg = ZDS_DSORG_PDSE;
+            entry.devtype = "3390";
+          }
+          else
+          {
+            // Extract the device type code from byte 3 (last byte)
+            uint8_t dev_code = static_cast<unsigned char>(data[3]);
+
+            // Map UCB device type to device name
+            // Reference: z/OS DFSMSdfp Advanced Services, UCB device type codes
+            switch (dev_code)
+            {
+            case 0x0B:
+              entry.devtype = "3340";
+              break;
+            case 0x0E:
+              entry.devtype = "3350";
+              break;
+            case 0x0F:
+              entry.devtype = "3390";
+              break; // Most common modern DASD
+            case 0x10:
+              entry.devtype = "9345";
+              break;
+            case 0x2E:
+              entry.devtype = "3380";
+              break;
```

Comment:
See comment above regarding tracks/cyl - we parse different device types, but hardcode 15 tracks/cyl for all space calculations. This could cause issues with 3340 and 3350